Set up C tests to fail upon warnings

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

In a couple of places we were using wrong string formatters (e.g. %d instead of %zu), sometimes the number of formatters didn't match the arguments, and sometimes the inline type annotation was too broad (int[] instead of uint32_t[]). 

Interestingly in #14095 this leads to CI failure, maybe something just change in a tool we were using? Anyways it's good to fix these warnings before they accumulate.

### Details and comments

Cryoris committed:
Fix string formatting and inaccurate types in C tests

In a couple of places we were using wrong string formatters (e.g. %d instead of %zu), sometimes the number of formatters didn't match the arguments, and sometimes the inline type annotation was too broad (int[] instead of uint32_t[]).

mtreinish reviewed (commented):
Do you want to make warnings fatal with this so we don't regress here in the future?

Cryoris committed:
Enable `-Wall -Wextra -Werror` to fail eagerly

... this did uncover some bugs

Cryoris commented:
Fatalities enabled 🙂 (and fought off some fatalities in the process lol)

Cryoris commented on crates/cext/include/complex.h:
```diff
@@ -29,28 +29,28 @@ typedef struct {
     return value->re + I * value->im;
 }
-static QkComplex64 qk_complex64_from_native(double _Complex *value) {
+static inline QkComplex64 qk_complex64_from_native(double _Complex *value) {
```
The compiler failed without this, though I'm not sure I fully understood why. As far as I understand it this header was included in some executables where we didn't actually call these functions, so the compiler complained about unused functions. Adding `inline` should avoid that since the function would not actually get included in executables where it isn't used.

Since these functions are so short it might not be bad to inline anyways?

mtreinish replied:
I don't think `inline` keyword should change that. The `inline` keyword is basically a hint to the compiler to optimize for speed for those function calls, so it might inline the code. It's meaning around a static function definition is a bit muddier to me though and it probably depends on the compiler's interpretation, I'd have to do more reading on it to be sure.

mtreinish replied:
Reading through the GCC docs on the `inline` keyword  it does seem appropriate here:

> When a function is both inline and static, if all calls to the function are integrated into the caller, and the function’s address is never used, then the function’s own assembler code is never referenced. In this case, GCC does not actually output assembler code for the function, unless you specify the option -fkeep-inline-functions. If there is a nonintegrated call, then the function is compiled to assembler code as usual. The function must also be compiled as usual if the program refers to its address, because that cannot be inlined.

Cryoris committed:
extra sausage for windows

Cryoris committed:
Merge branch 'main' into fix-c-compiler-warnings

Cryoris committed:
baptize the newcomers

Cryoris committed:
enable -Wshadow for MSVC consistency

sounds like a good thing to enable anyways

Cryoris committed:
downgrade MSVC to W3

W4 includes a bunch of pedantic warnings that are not included in GCCs -Wall -Wextra

Cryoris committed:
downgrade MSVC to W2

Cryoris committed:
missing void, more wrong formatters

Cryoris committed:
on second thought: remove MSVC only check on GCC

Cryoris commented:
After some discussion and thought: restricting the checks only to GCC gives a better defined standard that CI has to check. Including MSVC will always differ, so we remove `-Werror` there and commit only to passing it on GCC.

mtreinish commented on test/c/CMakeLists.txt:
```diff
@@ -3,6 +3,14 @@
 # https://cmake.org/cmake/help/book/mastering-cmake/chapter/Testing%20With%20CMake%20and%20CTest.html
 cmake_minimum_required (VERSION 3.27)
 
+# On GCC compilers, we fail upon any warnings. MSVC is not included since compiler warnings
```
What about clang? We only build with GCC on linux, on macOS we use Apple Clang to build the c tests. The CLI options here are the same on clang and gcc but the compilers are very different. I think you just need to update the comment, I'm not too worried about the differences between clang and gcc here that will give us broader coverage.

Cryoris replied:
Added a note on clang in dcdff2a 👍🏻

Cryoris committed:
note on clang
